# Data Loading

Importing libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### LOAD TRAIN

In [2]:
orders_train = pd.read_csv("../data/raw/train/df_Orders.csv")
order_items_train = pd.read_csv("../data/raw/train/df_OrderItems.csv")
payments_train = pd.read_csv("../data/raw/train/df_Payments.csv")
customers_train = pd.read_csv("../data/raw/train/df_Customers.csv")

### LOAD TEST

In [3]:
orders_test = pd.read_csv("../data/raw/test/df_Orders.csv")
order_items_test = pd.read_csv("../data/raw/test/df_OrderItems.csv")
payments_test = pd.read_csv("../data/raw/test/df_Payments.csv")
customers_test = pd.read_csv("../data/raw/test/df_Customers.csv")

### Merge Train + Test

In [4]:
orders = pd.concat([orders_train, orders_test], ignore_index=True)
order_items = pd.concat([order_items_train, order_items_test], ignore_index=True)
payments = pd.concat([payments_train, payments_test], ignore_index=True)
customers = pd.concat([customers_train, customers_test], ignore_index=True)

Checking for duplicates

In [5]:
print("Duplicates after merge:")
print("Orders:", orders.duplicated().sum())
print("Order Items:", order_items.duplicated().sum())
print("Payments:", payments.duplicated().sum())
print("Customers:", customers.duplicated().sum())

Duplicates after merge:
Orders: 0
Order Items: 0
Payments: 0
Customers: 0


checking the shape fo the dataset

train shape

In [6]:
print("Orders shape:", orders_train.shape)
print("Order Items shape:", order_items_train.shape)
print("Payments shape:", payments_train.shape)
print("Customers shape:", customers_train.shape)

Orders shape: (89316, 7)
Order Items shape: (89316, 5)
Payments shape: (89316, 5)
Customers shape: (89316, 4)


test shape

In [7]:
print("Orders shape:", orders_test.shape)
print("Order Items shape:", order_items_test.shape)
print("Payments shape:", payments_test.shape)
print("Customers shape:", customers_test.shape)

Orders shape: (38279, 4)
Order Items shape: (38279, 5)
Payments shape: (38279, 5)
Customers shape: (38279, 4)


after merge

In [8]:
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Payments:", payments.shape)
print("Customers:", customers.shape)

Orders: (127595, 7)
Order Items: (127595, 5)
Payments: (127595, 5)
Customers: (127595, 4)


looks good

89316 + 38279 = 127595

# Data Cleaning

In [9]:
display(orders.head())
display(order_items.head())
display(payments.head())
display(customers.head())


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_timestamp,order_estimated_delivery_date
0,Axfy13Hk4PIk,hCT0x9JiGXBQ,delivered,2017-10-22 18:57:54,2017-10-22 19:14:13,2017-10-26 22:19:52,2017-11-09
1,v6px92oS8cLG,PxA7fv9spyhx,delivered,2018-06-20 21:40:31,2018-06-20 22:20:20,2018-07-03 22:51:22,2018-07-24
2,Ulpf9skrhjfm,g3nXeJkGI0Qw,delivered,2018-02-16 16:19:31,2018-02-17 16:15:35,2018-02-27 01:29:50,2018-03-08
3,bwJVWupf2keN,EOEsCQ6QlpIg,delivered,2018-08-18 18:04:29,2018-08-18 18:15:16,2018-08-27 20:03:51,2018-09-19
4,Dd0QnrMk9Cj5,mVz5LO2Vd6cL,delivered,2017-12-22 16:44:04,2017-12-22 17:31:31,2018-01-05 19:22:49,2018-01-18


,order_id,product_id,seller_id,price,shipping_charges
0,Axfy13Hk4PIk,90K0C1fIyQUf,ZWM05J9LcBSF,223.51,84.65
1,v6px92oS8cLG,qejhpMGGVcsl,IjlpYfhUbRQs,170.80,23.79
2,Ulpf9skrhjfm,qUS5d2pEAyxJ,77p2EYxcM9MD,64.40,17.38
3,bwJVWupf2keN,639iGvMyv0De,jWzS0ayv9TGf,264.50,30.72
4,Dd0QnrMk9Cj5,1lycYGcsic2F,l1pYW6GBnPMr,779.90,30.66


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,Axfy13Hk4PIk,1,credit_card,1,259.14
1,v6px92oS8cLG,1,credit_card,8,382.39
2,Ulpf9skrhjfm,1,credit_card,4,249.25
3,bwJVWupf2keN,1,credit_card,2,27.79
4,Dd0QnrMk9Cj5,1,credit_card,1,76.15


,customer_id,customer_zip_code_prefix,customer_city,customer_state
0,hCT0x9JiGXBQ,58125,varzea paulista,SP
1,PxA7fv9spyhx,3112,armacao dos buzios,RJ
2,g3nXeJkGI0Qw,4119,jandira,SP
3,EOEsCQ6QlpIg,18212,uberlandia,MG
4,mVz5LO2Vd6cL,88868,ilhabela,SP


In [10]:
print("Orders info\n",orders.info())
print("Order Items info\n",order_items.info())
print("Payments info\n",payments.info())
print("Customers info\n",customers.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127595 entries, 0 to 127594
Data columns (total 7 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   order_id                       127595 non-null  object
 1   customer_id                    127595 non-null  object
 2   order_status                   89316 non-null   object
 3   order_purchase_timestamp       127595 non-null  object
 4   order_approved_at              127579 non-null  object
 5   order_delivered_timestamp      87427 non-null   object
 6   order_estimated_delivery_date  89316 non-null   object
dtypes: object(7)
memory usage: 6.8+ MB
Orders info
 None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127595 entries, 0 to 127594
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   order_id          127595 non-null  object 
 1   product_id        127595 non-null

### Orders tables Cleaning

Convert timestamps

In [11]:
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])
orders['order_approved_at'] = pd.to_datetime(orders['order_approved_at'])
orders['order_delivered_timestamp'] = pd.to_datetime(orders['order_delivered_timestamp'])
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'])

In [12]:
print(orders.isnull().sum())

order_id                             0
customer_id                          0
order_status                     38279
order_purchase_timestamp             0
order_approved_at                   16
order_delivered_timestamp        40168
order_estimated_delivery_date    38279
dtype: int64


Order Status Distribution

In [13]:
print("Order Status Counts:", orders['order_status'].value_counts())

Order Status Counts: order_status
delivered      87428
shipped          936
canceled         409
processing       273
invoiced         266
unavailable        2
approved           2
Name: count, dtype: int64


Checking Date Ranges

In [14]:
print("\nDate Range:")
print("Min:", orders['order_purchase_timestamp'].min())
print("Max:", orders['order_purchase_timestamp'].max())


Date Range:
Min: 2016-09-04 21:15:19
Max: 2018-09-03 09:06:57


### Insights

There are quite a few missing values
- order_status                     38279
- order_approved_at                   16
- order_delivered_timestamp        40168
- order_estimated_delivery_date    38279

But they might now be random, the orders might not have been dilivered therefore all this missing values are related to non delivered items and hence missing. This might not be usefull in our ML training because there we are trying to build a CLV model so only actual transactions which happened matters. but this might still matter in eda, so we will keep data aside in some form to do eda.

In [15]:
# Keep only delivered orders
orders_ml = orders[orders['order_status'] == 'delivered']

print("After filtering:", orders_ml.shape)

# Check missing again
print("\nMissing values after filtering:\n", orders_ml.isnull().sum())

After filtering: (87428, 7)

Missing values after filtering:
 order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                9
order_delivered_timestamp        6
order_estimated_delivery_date    0
dtype: int64


Saving for EDA

In [16]:
orders.to_csv("../data/processed/eda/orders_cleaned.csv", index=False)

cleaning rest missing values

In [17]:
orders_ml = orders_ml.dropna(subset=[
    'order_approved_at',
    'order_delivered_timestamp'
])

print("Final ML shape:", orders_ml.shape)
print("\nRemaining missing:\n", orders_ml.isnull().sum())

Final ML shape: (87413, 7)

Remaining missing:
 order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_timestamp        0
order_estimated_delivery_date    0
dtype: int64


In [18]:
orders_ml['order_status'] = orders_ml['order_status'].astype('category')

In [19]:
orders_ml['order_id'] = orders_ml['order_id'].astype('category')
orders_ml['customer_id'] = orders_ml['customer_id'].astype('category')

In [20]:
orders_ml.info()

<class 'pandas.core.frame.DataFrame'>
Index: 87413 entries, 0 to 89315
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       87413 non-null  category      
 1   customer_id                    87413 non-null  category      
 2   order_status                   87413 non-null  category      
 3   order_purchase_timestamp       87413 non-null  datetime64[ns]
 4   order_approved_at              87413 non-null  datetime64[ns]
 5   order_delivered_timestamp      87413 non-null  datetime64[ns]
 6   order_estimated_delivery_date  87413 non-null  datetime64[ns]
dtypes: category(3), datetime64[ns](4)
memory usage: 9.5 MB


### Order items table cleaning

In [21]:
print(order_items.isnull().sum())
print("\nDuplicates:", order_items.duplicated().sum())

# Check basic stats
print(order_items[['price', 'shipping_charges']].describe())

order_id            0
product_id          0
seller_id           0
price               0
shipping_charges    0
dtype: int64

Duplicates: 0
               price  shipping_charges
count  127595.000000     127595.000000
mean      341.380366         44.256468
std       558.307510         37.692326
min         0.850000          0.000000
25%        59.650000         20.020000
50%       136.900000         34.870000
75%       399.200000         57.160000
max      6735.000000        409.680000


In [22]:
order_items['order_id'] = order_items['order_id'].astype('category')
order_items['product_id'] = order_items['product_id'].astype('category')
order_items['seller_id'] = order_items['seller_id'].astype('category')

In [23]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127595 entries, 0 to 127594
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype   
---  ------            --------------   -----   
 0   order_id          127595 non-null  category
 1   product_id        127595 non-null  category
 2   seller_id         127595 non-null  category
 3   price             127595 non-null  float64 
 4   shipping_charges  127595 non-null  float64 
dtypes: category(3), float64(2)
memory usage: 9.5 MB


### Payments Table Cleaning

In [24]:
print(payments.isnull().sum())
print("\nDuplicates:", payments.duplicated().sum())

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Duplicates: 0


In [25]:
payments['order_id'] = payments['order_id'].astype('category')
payments['payment_type'] = payments['payment_type'].astype('category')

In [26]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,Axfy13Hk4PIk,1,credit_card,1,259.14
1,v6px92oS8cLG,1,credit_card,8,382.39
2,Ulpf9skrhjfm,1,credit_card,4,249.25
3,bwJVWupf2keN,1,credit_card,2,27.79
4,Dd0QnrMk9Cj5,1,credit_card,1,76.15


In [27]:
payments['payment_installments'] = payments['payment_installments'].astype('int16')

In [28]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127595 entries, 0 to 127594
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype   
---  ------                --------------   -----   
 0   order_id              127595 non-null  category
 1   payment_sequential    127595 non-null  int64   
 2   payment_type          127595 non-null  category
 3   payment_installments  127595 non-null  int16   
 4   payment_value         127595 non-null  float64 
dtypes: category(2), float64(1), int16(1), int64(1)
memory usage: 7.8 MB


ensuring for no negative values

In [29]:
print(payments['payment_value'].min())

0.0


### Customers Table CLeaning

In [30]:
print(customers.isnull().sum())
print("\nDuplicates:", customers.duplicated().sum())

customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Duplicates: 0


In [31]:
customers['customer_id'] = customers['customer_id'].astype('category')
customers['customer_city'] = customers['customer_city'].astype('category')
customers['customer_state'] = customers['customer_state'].astype('category')

In [32]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127595 entries, 0 to 127594
Data columns (total 4 columns):
 #   Column                    Non-Null Count   Dtype   
---  ------                    --------------   -----   
 0   customer_id               127595 non-null  category
 1   customer_zip_code_prefix  127595 non-null  int64   
 2   customer_city             127595 non-null  category
 3   customer_state            127595 non-null  category
dtypes: category(3), int64(1)
memory usage: 7.0 MB


In [33]:
customers.head()

,customer_id,customer_zip_code_prefix,customer_city,customer_state
0,hCT0x9JiGXBQ,58125,varzea paulista,SP
1,PxA7fv9spyhx,3112,armacao dos buzios,RJ
2,g3nXeJkGI0Qw,4119,jandira,SP
3,EOEsCQ6QlpIg,18212,uberlandia,MG
4,mVz5LO2Vd6cL,88868,ilhabela,SP


In [34]:
customers['customer_zip_code_prefix'] = customers['customer_zip_code_prefix'].astype('str')

# Data Integrity Check

### Key Integrity

In [35]:
# Orders ↔ Order Items
print("Order IDs missing in order_items:",
      orders_ml[~orders_ml['order_id'].isin(order_items['order_id'])].shape[0])

# Orders ↔ Payments
print("Order IDs missing in payments:",
      orders_ml[~orders_ml['order_id'].isin(payments['order_id'])].shape[0])

# Orders ↔ Customers
print("Customer IDs missing in customers:",
      orders_ml[~orders_ml['customer_id'].isin(customers['customer_id'])].shape[0])

Order IDs missing in order_items: 0
Order IDs missing in payments: 0
Customer IDs missing in customers: 0


### Revenue Consistency Check

In [36]:
# Merge small sample to check
temp = order_items.merge(payments, on='order_id')

temp['calc_total'] = temp['price'] + temp['shipping_charges']

print(temp[['calc_total', 'payment_value']].describe())

          calc_total  payment_value
count  127595.000000  127595.000000
mean      385.636834     268.369441
std       559.326771     351.122082
min         1.140000       0.000000
25%       107.720000      83.620000
50%       184.750000     171.860000
75%       437.860000     312.410000
max      6767.930000   13664.080000


### Timestamp Logic Check

In [37]:
# Approval before purchase?
print("Invalid approval:",
      (orders_ml['order_approved_at'] < orders_ml['order_purchase_timestamp']).sum())

# Delivery before purchase?
print("Invalid delivery:",
      (orders_ml['order_delivered_timestamp'] < orders_ml['order_purchase_timestamp']).sum())

Invalid approval: 0
Invalid delivery: 0


### Customer Behavior Check

In [38]:
orders_per_customer = orders_ml.groupby('customer_id')['order_id'].count()

print(orders_per_customer.describe())

count    87413.0
mean         1.0
std          0.0
min          1.0
25%          1.0
50%          1.0
75%          1.0
max          1.0
Name: order_id, dtype: float64


C:\Users\Administrator\AppData\Local\Temp\ipykernel_19932\2191878210.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  orders_per_customer = orders_ml.groupby('customer_id')['order_id'].count()


### Insight

count    87413.0

mean         1.0

std          0.0

min          1.0

25%          1.0

50%          1.0

75%          1.0

max          1.0

As we can see all th customer has 1 mean or 1 max. Meaning there are no repeated customers which is bas for CLV Approach. So we will now check for other insight to try different ML methodology. But first we will save the data.

# Data Cleaning part 2

First we didnt need Products table because we were doing clv, but now we are not doing CLV and trying different models. Product tables will be usefull now.

### Cleaning Products table

Loading Products Table

In [39]:
products_train = pd.read_csv("../data/raw/train/df_Products.csv")
products_test = pd.read_csv("../data/raw/test/df_Products.csv")

merging train + test

In [40]:
products = pd.concat([products_train, products_test], ignore_index=True)

In [41]:
products.head()

,product_id,product_category_name,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,90K0C1fIyQUf,toys,491.0,19.0,12.0,16.0
1,qejhpMGGVcsl,watches_gifts,440.0,18.0,14.0,17.0
2,qUS5d2pEAyxJ,costruction_tools_garden,2200.0,16.0,16.0,16.0
3,639iGvMyv0De,toys,1450.0,68.0,3.0,48.0
4,1lycYGcsic2F,toys,300.0,17.0,4.0,12.0


Checking null and duplicates

In [42]:
print(products.isnull().sum())
print("\nDuplicates:", products.duplicated().sum())

product_id                 0
product_category_name    476
product_weight_g          25
product_length_cm         25
product_height_cm         25
product_width_cm          25
dtype: int64

Duplicates: 94644


removing duplicates, only one row needed per product

In [43]:
products = products.drop_duplicates(subset='product_id')

In [44]:
print("After removing duplicates:", products.duplicated(subset='product_id').sum())
print("Shape:", products.shape)

After removing duplicates: 0
Shape: (32951, 6)


category missing

In [45]:
products['product_category_name'] = products['product_category_name'].fillna('unknown')

In [46]:
print(products.isnull().sum())

product_id               0
product_category_name    0
product_weight_g         2
product_length_cm        2
product_height_cm        2
product_width_cm         2
dtype: int64


Dimensions Missing

In [47]:
for col in ['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']:
    products[col] = products[col].fillna(products[col].median())
print(products.isnull().sum())

product_id               0
product_category_name    0
product_weight_g         0
product_length_cm        0
product_height_cm        0
product_width_cm         0
dtype: int64


In [48]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32951 entries, 0 to 127584
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   product_id             32951 non-null  object 
 1   product_category_name  32951 non-null  object 
 2   product_weight_g       32951 non-null  float64
 3   product_length_cm      32951 non-null  float64
 4   product_height_cm      32951 non-null  float64
 5   product_width_cm       32951 non-null  float64
dtypes: float64(4), object(2)
memory usage: 1.8+ MB


In [49]:
products.describe()

,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32951.000000,32951.000000,32951.000000,32951.000000
mean,2276.376802,30.814725,16.937422,23.196534
std,4281.926387,16.914005,13.637175,12.078707
min,0.000000,7.000000,2.000000,6.000000
25%,300.000000,18.000000,8.000000,15.000000
50%,700.000000,25.000000,13.000000,20.000000
75%,1900.000000,38.000000,21.000000,30.000000
max,40425.000000,105.000000,105.000000,118.000000


minimum weight 0 might be wrong

In [50]:
products[products['product_weight_g'] == 0].shape[0]

4

In [51]:
products[products['product_weight_g'] == 0]

,product_id,product_category_name,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1476,fsq5lo34nHGK,toys,0.0,30.0,25.0,30.0
51136,aMbBkxpim3xt,toys,0.0,30.0,25.0,30.0
65518,nm5AMMVOx4Ac,toys,0.0,30.0,25.0,30.0
83528,xUlGNJzxYn1N,toys,0.0,30.0,25.0,30.0


only 4 places we can replace them with median

In [52]:
products.shape

(32951, 6)

In [53]:
median_weight = products['product_weight_g'].median()

products['product_weight_g'] = products['product_weight_g'].replace(0, median_weight)

In [54]:
products[products['product_weight_g'] == 0].shape[0]

0

In [55]:
products['product_volume_cm3'] = (
    products['product_length_cm'] *
    products['product_height_cm'] *
    products['product_width_cm']
)

Convert categorical columns

In [56]:
products['product_id'] = products['product_id'].astype('category')
products['product_category_name'] = products['product_category_name'].astype('category')

optimization

In [57]:
for col in ['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']:
    products[col] = products[col].astype('float32')

# Saving Data

In [58]:
orders_ml.to_parquet("../data/processed/orders_ml_clean.parquet", index=False)
order_items.to_parquet("../data/processed/order_items_clean.parquet", index=False)
payments.to_parquet("../data/processed/payments_clean.parquet", index=False)
customers.to_parquet("../data/processed/customers_clean.parquet", index=False)
products.to_parquet("../data/processed/products_clean.parquet", index=False)

# ML Model Decision

In [59]:
print(payments['payment_value'].describe())

count    127595.000000
mean        268.369441
std         351.122082
min           0.000000
25%          83.620000
50%         171.860000
75%         312.410000
max       13664.080000
Name: payment_value, dtype: float64


In [60]:
print("Unique states:", customers['customer_state'].nunique())
print("Unique cities:", customers['customer_city'].nunique())

Unique states: 27
Unique cities: 4110


In [61]:
display(orders_ml.describe())
display(order_items.describe())
display(payments.describe())
display(customers.describe())

,order_purchase_timestamp,order_approved_at,order_delivered_timestamp,order_estimated_delivery_date
count,87413,87413,87413,87413
mean,2017-12-29 22:08:00.051639808,2017-12-30 08:30:16.404482048,2018-01-11 08:33:03.465502976,2018-01-22 18:57:26.837426944
min,2016-10-03 09:44:50,2016-10-04 09:43:32,2016-10-11 13:46:32,2016-10-27 00:00:00
25%,2017-09-10 20:15:46,2017-09-11 15:30:18,2017-09-20 20:27:47,2017-10-02 00:00:00
50%,2018-01-17 11:59:12,2018-01-17 16:35:21,2018-01-30 22:48:40,2018-02-14 00:00:00
75%,2018-05-02 23:06:51,2018-05-03 09:12:19,2018-05-14 15:54:50,2018-05-24 00:00:00
max,2018-08-29 14:52:00,2018-08-29 15:05:22,2018-10-17 13:22:46,2018-10-25 00:00:00


,price,shipping_charges
count,127595.000000,127595.000000
mean,341.380366,44.256468
std,558.307510,37.692326
min,0.850000,0.000000
25%,59.650000,20.020000
50%,136.900000,34.870000
75%,399.200000,57.160000
max,6735.000000,409.680000


,payment_sequential,payment_installments,payment_value
count,127595.000000,127595.000000,127595.000000
mean,1.098734,2.970634,268.369441
std,0.763641,2.808502,351.122082
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,83.620000
50%,1.000000,2.000000,171.860000
75%,1.000000,4.000000,312.410000
max,29.000000,24.000000,13664.080000


,customer_id,customer_zip_code_prefix,customer_city,customer_state
count,127595,127595,127595,127595
unique,127595,14976,4110,27
top,zzzEZ7FwuZqA,46550,sao paulo,SP
freq,1,280,20512,54067


As disscussed earlier that this dataset is weak for CLV, but from above insights it is a very good dataset for 
- Customer Value Prediction
- High Value Customer Classification
- Customer Segmentation

### Next --> EDA